pkg/list/lister.go

```diff
@@ -39,55 +49,122 @@ func NewLister(ctx context.Context) *lister {
 	}
 }
 
-func (l *lister) GetPackagesWithStatus(
+// same as NewLister, but stores the meta index after the first time it has been fetched – CAUTION: this cache
+// is not concurrency-safe and therefore the listers function should only be used one at a time
+func NewListerWithRepoCache(ctx context.Context) *lister {
+	l := NewLister(ctx)
+	l.useCache = true
+	return l
+}
+
+func (l *lister) GetClusterPackagesWithStatus(
 	ctx context.Context,
 	options ListOptions,
 ) ([]*PackageWithStatus, error) {
-	index, err := l.fetchRepoAndInstalled(ctx, options)
+	index, err := l.fetchRepoAndInstalled(ctx, options, includeClusterPackages)
 	result := make([]*PackageWithStatus, 0, len(index))
 	for _, item := range index {
-		pkgWithStatus := PackageWithStatus{
-			MetaIndexItem: *item.IndexItem,
+		if itemShouldBeIncluded(&item, options) {
+			pkgWithStatus := PackageWithStatus{
+				MetaIndexItem:  *item.IndexItem,
+				ClusterPackage: item.ClusterPackage,
+				Status:         client.GetStatusOrPending(item.ClusterPackage),
+			}
+			if item.PackageInfo != nil {
+				pkgWithStatus.InstalledManifest = item.PackageInfo.Status.Manifest
+			}
+			result = append(result, &pkgWithStatus)
 		}
+	}
+	return result, err
+}
 
-		if !((options.OnlyInstalled && !item.Installed()) || (options.OnlyOutdated && !item.Outdated())) {
-			pkgWithStatus.Package = item.Package
-			pkgWithStatus.Status = client.GetStatusOrPending(item.Package)
-
+func (l *lister) GetPackagesWithStatus(
+	ctx context.Context,
+	options ListOptions,
+) ([]*PackagesWithStatus, error) {
+	index, err := l.fetchRepoAndInstalled(ctx, options, includePackages)
+	result := make([]*PackagesWithStatus, 0, len(index))
+	for _, item := range index {
+		// TODO itemShouldBeIncluded is wrong here – need to check outdated in the inner loop
+		// if itemShouldBeIncluded(&item, options) {
+		ls := make([]*PackageWithStatus, 0, len(item.Packages))
+		for _, pkg := range item.Packages {
+			pkgWithStatus := PackageWithStatus{
+				MetaIndexItem: *item.IndexItem,
+				Package:       pkg,
+				Status:        client.GetStatusOrPending(pkg),
+			}
 			if item.PackageInfo != nil {
 				pkgWithStatus.InstalledManifest = item.PackageInfo.Status.Manifest
 			}
-
-			result = append(result, &pkgWithStatus)
+			ls = append(ls, &pkgWithStatus)
 		}
+		result = append(result, &PackagesWithStatus{
+			MetaIndexItem: *item.IndexItem,
+			Packages:      ls,
+		})
+		// }
 	}
 	return result, err
 }
 
-func (l *lister) fetchRepoAndInstalled(ctx context.Context, options ListOptions) (
+func itemShouldBeIncluded(item *result, options ListOptions) bool {
+	return !((options.OnlyInstalled && !item.Installed()) || (options.OnlyOutdated && !item.Outdated()))
+}
+
+type typeOptions int
+
+const (
+	// TODO is this how this works?
+	includeClusterPackages typeOptions = 1 << iota
+	includePackages
+	includeAll = 3
```

Comment:
Regarding the TODO comment: Yes, this will set the first const to 1, the second to 2, the third to 4, etc.

I would change the last item to make it more explicit and give it the correct type :slightly_smiling_face: 
```go
includeAll = includeClusterPackages | includePackages
```